Fixes #10558 - adds migration to remove and prevent duplicate tokens

orrabin opened the pull request.

orrabin committed:
Fixes #10558 - adds migration to remove and prevent duplicate tokens

dLobatog commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,18 @@
+      Token.where('id not in (?)', existing_tokens).delete_all
+    end
+
+    remove_index :tokens, :host_id
```
I think this is failing on MySQL if it's trying to remote an index that might not exist.

orrabin replied:
thanks, I'll try to check if it exists but it looks like it would happen even if the index exists `Mysql2::Error: Cannot drop index 'index_tokens_on_host_id': needed in a foreign key constraint`

dLobatog commented:
Thanks @orrabin, I'm curious as to how to properly test this myself?

orrabin commented:
@eLobato  you can crate the duplicate tokens through discovery prior to https://github.com/theforeman/foreman/commit/c2ff4bd6874155459a51af070abb381fd59be262 and then try to run this migration.

domcleal commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,18 @@
+      Token.where('id not in (?)', existing_tokens).delete_all
+    end
+
+    remove_index :tokens, :host_id if index_exists? :tokens, :host_id
```
I think you might need to temporarily remove the foreign key between hosts and tokens in order to do this under MySQL... also check if the foreign key exists first!

domcleal commented:
Another thing to consider, along the lines of the optimisation that was looked at in Discovery's version, is to get the list of duplicate host_ids instead of the lists of active tokens, then loop through it:

```
Token.having('count(*) > 1').group(:host_id).select('host_id')
```

Then the expense is only for the hosts and cases where the error exists, instead of generating a large query for all active-build hosts.

domcleal commented:
I see the PR's been updated, but any reply to my last comment?

orrabin commented:
@domcleal your solution is better for the duplicate tokens but in discovery there are also tokens created when trying to provision a discovered host and then canceling. If there is even one token for a host that isn't managed it should also be deleted.
Is it better to solve both problems here and remove the discovery migration or handle only duplicates here and keep the original migration in discovery?

domcleal commented:
Wasn't that from a different bug in Discovery itself?  I'm struggling to navigate the bug numbers to find it though.  If it's different we should probably keep both doing different things.

orrabin commented:
@domcleal no it's the same source of bug, the issue was moved from discovery to forman - `set_token` using create and not build (and not checking if a token exists already) caused an option to create multiple tokens through discovery, then provision and create another token.
We thought it was a discovery bug and tried to fix it there but then found the root cause was actually the set_token.
These issues are actually dups (or symptom and cause):
http://projects.theforeman.org/issues/9597
http://projects.theforeman.org/issues/10535

So if you have a managed host with more then one token it probably used to be a discovered host with a token that shouldn't have been saved.

domcleal commented:
Ah okay, then yeah, we should address it here and then it's up to you whether to keep the discovery migration (probably not, or change it to no-op).

Is it possible to do a join for tokens against hosts, where hosts.type != 'Host::Managed', and destroy those?

orrabin commented:
@domcleal yes, `Token.joins(:host)` returns only tokens connected to managed hosts so I can do this in 2 steps:
1) delete all tokens not in `Token.joins(:host)`
2) find duplicate tokens for managed hosts and delete them

domcleal commented:
> 1) delete all tokens not in Token.joins(:host)

My only thought is that this might be more expensive than trying to find the bad entries, at least on a regular installation where nothing's wrong.  If it's possible to search for tokens that are associated to non-Host::Managed hosts, then we might be able to make it more efficient.

orrabin commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,27 @@
+class RemoveDuplicateTokens < ActiveRecord::Migration
+  def up
+    # only managed hosts should have tokens
+    Token.where(:id => Token.joins("left outer join hosts on hosts.id = tokens.host_id").where("hosts.type != 'Host::Managed'").pluck(:id)).delete_all
```
`Token.joins("left outer join hosts on hosts.id = tokens.host_id").where("hosts.type != 'Host::Managed'").delete_all` resulted in an error:

```
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "hosts"
LINE 1: DELETE FROM "tokens" WHERE (hosts.type != 'Host::Managed')
```

ideas on how to delete these records without using the tokens table twice?

domcleal replied:
No, I don't.. but I think this is probably OK.

Edit: provided it runs as a subquery.. will check.

domcleal replied:
Do change `pluck(:id)` to `hosts.id` as well.

ShimShtein replied:
+1 @domcleal, change `pluck(:id)` to `select(:id)`. AFAIK most SQL servers do not allow delete with join syntax (hence subquery is more generic solution).

orrabin replied:
@ShimShtein thanks for checking, I'll change to `select('tokens.id')`

domcleal commented:
Note that this is WoC due to test failures.

orrabin commented:
@domcleal  thanks.
Apparently mysql doesn't like updates/deletes with subselect: http://dasunhegoda.com/1093-you-cant-specify-target-table-table_name-for-update-in-from-clause/104/
I changed back to `pluck`

orrabin commented:
The other option I thought about was keeping the subselect, selecting all of the token's ids and then deleting them:

``` ruby
tokens_to_delete = Token.where(:id => Token.joins("left outer join hosts on hosts.id = tokens.host_id").
                                 where("hosts.type != 'Host::Managed'").select('tokens.id')).pluck(:id)
Tokens.where(:id => tokens_to_delete).delete_all
```

I'm not sure which is worse. Any ideas?

domcleal commented:
> The other option I thought about was keeping the subselect, selecting all of the token's ids and then deleting them:
> 
> tokens_to_delete = Token.where(:id => Token.joins("left outer join hosts on hosts.id = tokens.host_id").
>                                  where("hosts.type != 'Host::Managed'").select('tokens.id')).pluck(:id)
> Tokens.where(:id => tokens_to_delete).delete_all

This looks identical to what's there right now on lines 4-5, where it gets all token IDs for non-Host::Managed and deletes them?  Just one line versus two.

No worries about the MySQL problem, we'll have to live with this, thanks for trying it.

domcleal commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,28 @@
+
+    # deleting duplicate tokens
+    hosts_with_duplicate_tokens = Token.having('count(*) > 1').group(:host_id).select('host_id')
+    existing_tokens = hosts_with_duplicate_tokens.map(&:token).compact
```
This doesn't look quite right to me (untested right now), as `hosts_with_duplicate_tokens` will be an array of host ID ints, but this line (9) is then calling .token on all those ints - they won't be Host objects AFAICT.

Perhaps:

```
hosts_with_duplicate_tokens.each do |host_id|
  Token.where(:host_id => host_id).offset(1).destroy_all
end
```

orrabin replied:
@domcleal using `offset(1)` means I assume the first token returned in the correct one. I don't mind doing this if we don't care which token is kept as long as there is only one but maybe I should change to `Host.where(:id => hosts_with_duplicate).map(&:token)`?

domcleal replied:
Yeah, I don't know which is going to be more correct.. it's probably going to depend a lot on how consistent the SQL query responses are, which varies depending on the server.  It might be they're both inaccurate if the SQL server changes the ordering.

orrabin commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,28 @@
+
+    # deleting duplicate tokens
+    hosts_with_duplicate_tokens = Token.having('count(*) > 1').group(:host_id).select('host_id')
+    existing_tokens = Host.where(:id => hosts_with_duplicate_tokens).map(&:token).compact
```
@domcleal I changed this to at least be consistent with `host.token`

lzap commented:
Ok so let me create a wrap-up here. Relevant issues:
- http://projects.theforeman.org/issues/9597 - original migration attempt (discovery)
- http://projects.theforeman.org/issues/10365 - fix for above 9597 (discovery)
- http://projects.theforeman.org/issues/10535 - fixed real cause (foreman)
- http://projects.theforeman.org/issues/10558 - this issue (to move migrations to foreman core)

Once this is merged, we will remove 9597 and 10365 from develop branch of discovery (future 4.0 version). Dominic, is your intention to backport this patch into 1.8? If not, we will make sure the above two patches are in 3.0 because this version is still affected by this bug (it only contains 9597 patch at the moment which does not fix all the cases).

domcleal commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,28 @@
+    if existing_tokens.empty?
+      Token.delete_all
+    else
+      Token.where('id not in (?)', existing_tokens).delete_all
```
`existing_tokens` is a problem, as it isn't _all_ tokens.  It's only a list of tokens for hosts with duplicate tokens, so hosts with a single, valid token aren't in this list, so get deleted...?

domcleal replied:
Perhaps say "where id not in (existing_tokens) and host_id in (hosts_with_duplicate_tokens)"?

Edit: though you'd probably need to get rid of or change that delete_all branch above?

domcleal commented:
> Dominic, is your intention to backport this patch into 1.8? If not, we will make sure the above two patches are in 3.0 because this version is still affected by this bug (it only contains 9597 patch at the moment which does not fix all the cases).

Yeah, I don't think this will be in 1.8 at this stage.  1.8.2 included the validation fix that prevents any new tokens being created, but not this migration to fix the mess.

lzap commented:
I am not sure about @orrabin availability after Summit. If this blocks 1.9 release I can take over from here. Let me know.

domcleal commented:
Not particularly, don't worry.

orrabin commented:
@domcleal I'm sorry it took me so long to respond, I rebased and changed the `delete_all` in `existing_tokens`.
I wasn't sure where you wanted me to change the other `delete_all`? in this PR or the discovery one?

domcleal commented on db/migrate/20150525081931_remove_duplicate_tokens.rb:
```diff
@@ -0,0 +1,28 @@
+    if existing_tokens.empty?
+      Token.where(:id => Token.where(:host_id => hosts_with_duplicate_tokens).pluck('tokens.id')).delete_all
+    else
+      Token.where(:id => Token.where('host_id in (?) and id not in (?)', hosts_with_duplicate_tokens, existing_tokens).pluck('tokens.id')).delete_all
```
I see a problem hosts_with_duplicate_tokens here, and wonder if the select('host_id') on line 8 needs to be a pluck.  This line generates a query such as:

```
SELECT tokens.id FROM "tokens" WHERE (host_id in (NULL,NULL,NULL,NULL,...) and id not in (50007,50012,...)
```

orrabin replied:
@domcleal I didn't see this but I changed to pluck and rebased

domcleal commented:
@orrabin I think it was a delete_all in this PR, but you've fixed it now.

domcleal commented:
Merged as 56608fe, thanks @orrabin.

lzap commented:
@orrabin we will be releasing Discovery 4.0 with both https://github.com/theforeman/foreman_discovery/pull/166/files and https://github.com/theforeman/foreman_discovery/pull/179/files and for develop/5.0 I think we can keep it as well since they will be both already executed.
